lib/evaluate/characteristics.h

```diff
@@ -129,6 +130,9 @@ struct DummyDataObject {
   explicit DummyDataObject(TypeAndShape &&t) : type{std::move(t)} {}
   explicit DummyDataObject(DynamicType t) : type{t} {}
   bool operator==(const DummyDataObject &) const;
+  bool operator!=(const DummyDataObject &that) const {
+    return !(*this == that);
+  }
```

Comment:
Is it possible to write a generic `!=` non-member function and avoid the need for all of these member functions?
```
template<typename T> bool operator!=(const T &x, const T &y) {
  return !(x == y);
}
```

Reply:
I don't know.